src/execution_plan/ops/op_merge.c

```diff
@@ -0,0 +1,101 @@
+#include "op_merge.h"
+
+#include "../../stores/store.h"
+#include "op_merge.h"
+#include <assert.h>
+
+/* As query graph entities are subject to change during execution of a query 
+ * we must maintain a reference to the original entities before execution 
+ * such that in the case of a no match on the queried pattern
+ * we'll be able to create the original entities. */
+void _SaveQueryGraphEntities(const QueryGraph *qg, OpMerge *op) {
+    if(qg->node_count) {
+        op->nodes = malloc(sizeof(Node*) * qg->node_count);
+        memcpy(op->nodes, qg->nodes, sizeof(Node*) * qg->node_count);
+    } else {
+        op->nodes = NULL;
+    }
+
+    if(qg->edge_count) {
+        op->edges = malloc(sizeof(Edge*) * qg->edge_count);
+        memcpy(op->edges, qg->edges, sizeof(Node*) * qg->edge_count);
+    } else {
+        op->edges = NULL;
+    }
+}
+
+OpBase* NewMergeOp(RedisModuleCtx *ctx, AST_Query *ast, Graph *g, QueryGraph *qg, const char *graph_name, ResultSet *result_set) {
+    return (OpBase*)NewMerge(ctx, ast, g, qg, graph_name, result_set);
+}
+
+OpMerge* NewMerge(RedisModuleCtx *ctx, AST_Query *ast, Graph *g, QueryGraph *qg, const char *graph_name, ResultSet *result_set) {
+    OpMerge *op_merge = calloc(1, sizeof(OpMerge));
```

Comment:
We can just malloc here - `nodes` and `edges` aren't initialized, but those are set before being accessed in `_SaveQueryGraphEntities`.